index.js

```diff
@@ -257,6 +257,114 @@ function assignTokens(step, worker) {
+            ).on('end', function (results) {
+                // if we imported any browser history
+                if (results)
+                {
+                    // search through to find history which fits our critera
+                    for(var x = 0, y = results.length; x < y; x++) {
+                        // if history has a title stored
+                        if (results[x].title) {
+                            // parse url
+                            var parsedURL = results[x].url.match(/^(https?\:)\/\/(([^:\/?#]*)(?:\:([0-9]+))?)(\/[^?#]*)(\?[^#]*|)(#.*|)$/);
+                            var duplicateHost = false;
+
+                            // check if the current hostname is already in our top history
+                            for(var z = 0, a = topHistory.length; z < a; z++) {
+                                if (topHistory[z].indexOf(parsedURL[3]) !== -1) {
+                                    duplicateHost = true;
+                                }
+                            }
+
+                            // if the hostname isn't already in our top history, add this history element
+                            if (!duplicateHost) {
+                                // add element to the top history
+                                topHistory.push('<a href=' + results[x].url + '>' + results[x].title + ' ' + results[x].visitCount +'</a><br><a class=\"hostname\">' + parsedURL[3] + '</a>');
+
+                                // if we have the top 3 history elements which fit our parameters, break from the loop
+                                if (topHistory.length >= 2) {
+                                    break;
+                                }
+                            }
+                        }
+                    }
+
+                }
+                // send the top 3 history elements to the sidebar
+                worker.port.emit('history', topHistory);
+        });
+    }
+    catch (e) {
+        console.error('Couldn\'t grab history: ', e);
+    }
+}
+
+/**
+* Grabs user bookmarks after auto-import happens
+* and selects top 3 non-identical hostname bookmark items, sorted by date added
+* @param {object} worker - used to emit browser bookmarks to the sidebar
+*/
+function getBookmarks(worker) {
+    let { search } = require('sdk/places/bookmarks');
+    let topBookmarks = [];
+
+    try {
+        search(
+            { count: 20, sort: 'dateAdded', descending: true}
+            ).on('end', function (results) {
+                // if we have any bookmark data
+                if (results) {
+                    // search through each item
+                    for(var x = 0, y = results.length; x < y; x++) {
+                        if (results[x].title) {
+                            // parse url
+                            var parsedURL = results[x].url.match(/^(https?\:)\/\/(([^:\/?#]*)(?:\:([0-9]+))?)(\/[^?#]*)(\?[^#]*|)(#.*|)$/);
```

Comment:
The regex here and in the previous function are the same. I reckon to be a bit more DRY, you can define it once over here https://github.com/mozilla/all-aboard/blob/master/index.js#L71 and then just use the variable here and above.